### PR TITLE
New release with fix

### PR DESCRIPTION
**Pull request by VatsalJagani**

### CODE REVIEW

File: appserver/static/mmdb_configuration.js [hunk 1]
```diff
@@ -30,9 +32,19 @@ require([
...
+        let msg_location = "#mmdb_configuration_mmdb_license_key_msg";
+        $(msg_location).text(` `);
+
         let maxmind_database_account_id = $('#mmdb_configuration_mmdb_account_id').val();
         let maxmind_database_license_key = $("#mmdb_configuration_mmdb_license_key").val();
-        let msg_location = "#mmdb_configuration_mmdb_license_key_msg";
+
+        let mmdb_config_proxy_url = $("#mmdb_config_proxy_url").val();
+        if (mmdb_config_proxy_url.trim() == ""){
+            mmdb_config_proxy_url = "None";
```
Comment (mahirchavda) [inline appserver/static/mmdb_configuration.js:43]: We should use null directly.

File: appserver/static/mmdb_configuration.js [hunk 2]
```diff
@@ -42,7 +54,9 @@ require([
         let service = mvc.createService();
         let data = {
             "maxmind_database_account_id": maxmind_database_account_id,
-            "maxmind_database_license_key": maxmind_database_license_key
+            "maxmind_database_license_key": maxmind_database_license_key,
+            "mmdb_config_proxy_url": mmdb_config_proxy_url
+            // "mmdb_config_is_ssl_verify": mmdb_config_is_ssl_verify
```
Comment (mahirchavda) [inline appserver/static/mmdb_configuration.js:59]: Please remove if not needed.

File: bin/rh_maxmind_configuration.py [hunk 3]
```diff
@@ -47,12 +47,24 @@ def handleList(self, conf_info):
...
+
             for i in data:
                 if i['name'] == mmdb_utils.MMDB_CONF_STANZA:
                     account_id = i['content']['account_id']
+                    # is_ssl_verify = i['content']['is_ssl_verify']
                     break
             conf_info['action']['maxmind_database_account_id'] = account_id
             conf_info['action']['maxmind_database_license_key'] = license_key
+            conf_info['action']['mmdb_config_proxy_url'] = mmdb_config_proxy_url
+            # conf_info['action']['mmdb_config_is_ssl_verify'] = is_ssl_verify
```
Comment (mahirchavda) [inline bin/rh_maxmind_configuration.py:67]: can we remove all commented code if not required.

File: bin/mmdb_utils.py [hunk 4]
```diff
@@ -241,23 +284,31 @@ def update_lookup(self):
...
-        '''
-        proxies = {
-            "http" : "<proxy-supported-schema http|https>://<username>:<password>@<ip-address>:<port>",
-            "https" : "<proxy-supported-schema http|https>://<username>:<password>@<ip-address>:<port>"
-        }
-        '''
+        if proxy_url:
+            if proxy_url.startswith("https"):
+                proxies = {
+                    "https" : proxy_url
```
Comment (mahirchavda) [inline bin/mmdb_utils.py:292]: I think we need to add both http and https with same proxy_url. Not 100% sure.

File: bin/mmdb_utils.py [hunk 5]
```diff
@@ -47,6 +60,15 @@
 
 
 
+def convert_to_bool_default_true(val):
+    if not val:
```
Comment (mahirchavda) [inline bin/mmdb_utils.py:64]: do we need if? it will be true by default

File: bin/mmdb_utils.py (hunk 4 above)
Comment (hardikhdholariya) [reply bin/mmdb_utils.py:292]: Yes, requires both in every case, it uses https or http values based on the target URL schema.

File: bin/mmdb_utils.py [hunk 6]
```diff
@@ -129,21 +151,39 @@ def __init__(self, session_key):
...
+        else:
+            logger.info("Using proxy_url provided by the user.")
+
+        # SSL certificate validation
+        is_ssl_verify = convert_to_bool_default_true(mmdb_config['is_ssl_verify'])
+        logger.info("Max Mind is_ssl_verify={}".format(is_ssl_verify))
+
         # Download mmdb database in a appropriate directory
-        self.download_mmdb_database(account_id, license_key)
+        self.download_mmdb_database(mmdb_config['account_id'], license_key, proxy_url, True)
```
Comment (hardikhdholariya) [inline bin/mmdb_utils.py:186]: need to pass is_ssl_verify instead of True

File: bin/mmdb_utils.py (hunk 4 above)
Comment (VatsalJagani) [reply bin/mmdb_utils.py]: Done

File: bin/mmdb_utils.py (hunk 5 above)
Comment (VatsalJagani) [reply bin/mmdb_utils.py]: Done

File: bin/mmdb_utils.py (hunk 6 above)
Comment (VatsalJagani) [reply bin/mmdb_utils.py:178]: nice catch, thanks!! Resolved.

Comment (VatsalJagani) [reply bin/mmdb_utils.py]: nice catch, thanks. Resolved.

**Review by mahirchavda — APPROVED** [review]

**Review by mahirchavda — APPROVED** [review]

File: bin/mmdb_utils.py [hunk 7]
```diff
@@ -129,24 +138,47 @@ def __init__(self, session_key):
...
+            msg = "Max Mind license key not found in password store. Please update config from Max Mind database configuration page."
             logger.error(msg)
             raise Exception(msg)
 
+        # Read MaxMind license key
+        proxy_url = None
+        try:
+            proxy_url = self.get_max_mind_proxy_url()
+        except Exception as e:
+            logger.info(f"Exception in getting proxy_url. {f}")
```
Comment (hardikhdholariya) [inline bin/mmdb_utils.py:159]: {e} instead of {f}

Comment (VatsalJagani) [reply bin/mmdb_utils.py]: Thanks for capturing it. Fixed it.

File: bin/mmdb_utils.py [hunk 8]
```diff
@@ -129,24 +138,47 @@ def __init__(self, session_key):
...
 
         # Read MaxMind license key
         license_key = self.get_max_mind_license_key()
         if not license_key:
-            msg = "Max Mind license key not found in password store. Please update config from Max Mind database configuration page.."
+            msg = "Max Mind license key not found in password store. Please update config from Max Mind database configuration page."
             logger.error(msg)
             raise Exception(msg)
 
+        # Read MaxMind license key
```
Comment (hardikhdholariya) [inline bin/mmdb_utils.py:154]: Read Proxy URL

Comment (VatsalJagani) [reply bin/mmdb_utils.py]: Fixed.

File: bin/mmdb_utils.py [hunk 9]
```diff
@@ -129,24 +138,47 @@ def __init__(self, session_key):
...
         if not license_key:
-            msg = "Max Mind license key not found in password store. Please update config from Max Mind database configuration page.."
+            msg = "Max Mind license key not found in password store. Please update config from Max Mind database configuration page."
             logger.error(msg)
             raise Exception(msg)
 
+        # Read Proxy URL
+        proxy_url = None
+        try:
+            proxy_url = self.get_max_mind_proxy_url()
```
Comment (hardikhdholariya) [inline bin/mmdb_utils.py:157]: Have you tested with the proxy password having special characters in it? if it has @ then it should be passed in encoded format.

Comment (VatsalJagani) [reply bin/mmdb_utils.py:177]: Thanks for catching this. 

I have not tested it with real proxy but I have tested the functions with below combinations

```
print(encode_username_password_in_proxy_url("https://1.1.1.1"))
print(encode_username_password_in_proxy_url("https://1.1.1.1:8888"))
print(encode_username_password_in_proxy_url("https://abc:xyz@1.1.1.1:8888"))
print(encode_username_password_in_proxy_url("https://abc:hy%rsz*n@1.1.1.1:8888"))
print(encode_username_password_in_proxy_url("https://sbc:vatsal@123@1.1.1.1:8888"))
print(encode_username_password_in_proxy_url("https://vatsal@crossrealms.com:vatsal@123@1.1.1.1:8888"))
```

Answers

```
https://1.1.1.1
https://1.1.1.1:8888
https://abc:xyz@1.1.1.1:8888
https://abc:hy%25rsz%2An@1.1.1.1:8888
https://sbc:vatsal%40123@1.1.1.1:8888
https://vatsal%40crossrealms.com:vatsal%40123@1.1.1.1:8888
```

Comment (hardikhdholariya) [reply bin/mmdb_utils.py:177]: Just check that urlparse supports the socks4 and socks5 scheme extraction. rest changes are good.

Comment (hardikhdholariya) [reply bin/mmdb_utils.py:177]: As per the doc, it supports only these schemes: https://docs.python.org/3/library/urllib.parse.html
image

Comment (VatsalJagani) [reply bin/mmdb_utils.py:177]: Will look into it and try to solve it some other way.

Comment (hardikhdholariya) [reply bin/mmdb_utils.py:181]: Assuming user won't have colon (:) in it name itself.  Rest of all cases are covered and looks good.

**Review by hardikhdholariya — APPROVED** [review]

**Pull request merged**